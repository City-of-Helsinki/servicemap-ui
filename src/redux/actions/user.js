import LocalStorageUtility from '../../utils/localStorage';
import fetchAddress from '../../views/MapView/utils/fetchAddress';

export const setLocale = locale => ({
  type: 'SET_LOCALE',
  locale,
});

export const setInitialLoad = () => ({
  type: 'SET_INITIAL_LOAD',
});

export const setPage = page => ({
  type: 'SET_CURRENT_PAGE',
  page,
});

export const setUserPosition = position => ({
  type: 'SET_POSITION',
  position,
});

const setTheme = theme => ({
  type: 'SET_THEME',
  theme,
});


export const changeLocaleAction = locale => async (dispatch) => {
  dispatch(setLocale(locale));
};

export const actionSetInitialLoad = () => async (dispatch) => {
  dispatch(setInitialLoad());
};

export const setCurrentPage = page => async (dispatch) => {
  dispatch(setPage(page));
};

export const changeTheme = theme => async (dispatch) => {
  LocalStorageUtility.saveItem('theme', theme);
  dispatch(setTheme(theme));
};

export const findUserLocation = () => async (dispatch) => {
  const success = (position) => {
    if (position.coords.accuracy < 10000) {
      fetchAddress({ lat: position.coords.latitude, lng: position.coords.longitude })
        .then((data) => {
          dispatch(setUserPosition({
            coordinates: position.coords,
            allowed: true,
            addressData: data,
          }));
        });
    } else {
      console.warn(`Position accuracy: ${position.coords.accuracy}. Max accuracy: 10000`);
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  const error = (err) => {
    console.warn(`GeoLocation error:(${err.code}): ${err.message}`);
    if (err.code === 1) {
      dispatch(setUserPosition({ coordinates: null, allowed: false, addressData: null }));
    } else {
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  navigator.geolocation.getCurrentPosition(success, error);
};


export default { setLocale, changeLocaleAction };
