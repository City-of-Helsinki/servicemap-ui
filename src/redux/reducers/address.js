const initialState = {
  units: [],
  addressCoordinates: null,
  addressData: null,
  adminDistricts: [],
  toRender: 'adminDistricts',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_ADDRESS_DATA':
      return {
        ...state,
        addressData: action.data,
      };
    case 'SET_ADDRESS_LOCATION':
      return {
        ...state,
        addressCoordinates: action.location.addressCoordinates,
      };
    case 'SET_ADDRESS_UNITS':
      return {
        ...state,
        units: action.units,
      };
    case 'SET_ADMINISTRATIVE_DISTRICTS':
      return {
        ...state,
        adminDistricts: action.data,
      };
    case 'SET_TO_RENDER':
      return {
        ...state,
        toRender: action.data,
      };
    default:
      return state;
  }
};
