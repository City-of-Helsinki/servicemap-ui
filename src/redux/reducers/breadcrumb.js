const initialState = [];

const breadcrumbReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'PUSH_ENTRY':
      return [...state, action.entry];
    case 'POP_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return newEntries;
    }
    case 'REPLACE_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return [...newEntries, action.entry];
    }
    default:
      return state;
  }
};

export default breadcrumbReducer;
