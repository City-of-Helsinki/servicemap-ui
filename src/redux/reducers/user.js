
const initialState = {
  initialLoad: false,
  locale: 'fi',
  page: 'home',
  theme: 'default',
  position: {
    coordinates: null,
    allowed: false,
  },
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_LOCALE':
      return {
        ...state,
        locale: action.locale,
      };
    case 'SET_INITIAL_LOAD':
      return {
        ...state,
        initialLoad: true,
      };
    case 'SET_CURRENT_PAGE':
      return {
        ...state,
        page: action.page,
      };
    case 'SET_POSITION':
      return {
        ...state,
        position: action.position,
      };
    case 'SET_THEME':
      return {
        ...state,
        theme: action.theme,
      };
    default:
      return state;
  }
};
