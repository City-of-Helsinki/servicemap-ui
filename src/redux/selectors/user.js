
export const getTheme = state => state.user.theme;
export const getPage = state => state.user.page;
export const getLocale = state => state.user.locale;
