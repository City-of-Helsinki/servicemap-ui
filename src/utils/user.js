import { useSelector } from 'react-redux'
 
export const useUserTheme = () => useSelector(state => state.user.theme);
export const useUserLocale = () => useSelector(state => state.user.locale);
