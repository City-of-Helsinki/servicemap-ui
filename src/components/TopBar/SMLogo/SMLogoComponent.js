import { ButtonBase, NoSsr } from '@mui/material';
import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import { useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import HomeLogo from '../../Logos/HomeLogo';
import { isHomePage } from '../../../utils/path';

const SMLogoComponent = ({
  onClick,
  classes,
}) => {
  const intl = useIntl();
  const theme = useSelector(state => state.user.theme);
  const location = useLocation();
  const isOnHomePage = isHomePage(location?.pathname);
  return (
    <ButtonBase aria-current={isOnHomePage ? 'page' : null} aria-label={intl.formatMessage({ id: 'general.home.logo.ariaLabel' })} role="link" onClick={onClick}>
      <NoSsr>
        <HomeLogo aria-hidden contrast={theme === 'dark'} className={classes.logo} />
      </NoSsr>
    </ButtonBase>
  );
};

SMLogoComponent.propTypes = {
  onClick: PropTypes.func.isRequired,
  classes: PropTypes.shape({
    logo: PropTypes.string,
  }).isRequired,
};

export default SMLogoComponent;
