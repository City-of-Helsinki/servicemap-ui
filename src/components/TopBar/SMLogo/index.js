import { withStyles } from '@mui/styles';
import SMLogoComponent from './SMLogoComponent';
import styles from '../styles';

const SMLogo = withStyles(styles)(SMLogoComponent);

export default SMLogo;
