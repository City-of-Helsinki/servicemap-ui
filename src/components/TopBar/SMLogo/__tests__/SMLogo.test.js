// Link.react.test.js
import React from 'react';
import { fireEvent } from '@testing-library/react';
import SMLogo from '../index';
import { initialState } from '../../../../redux/reducers/user';
import { getRenderWithProviders } from '../../../../../jestUtils';
import englishTranslations from '../../../../i18n/en';

const renderWithProviders = getRenderWithProviders({
  user: initialState,
  settings: {},
});

describe('<SMLogo />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<SMLogo />);
    expect(container).toMatchSnapshot();
  });

  it('simulates click event', () => {
    const mockCallBack = jest.fn();
    const { container } = renderWithProviders(
      <SMLogo
        onClick={mockCallBack}
      />,
    );
    fireEvent.click(container.querySelector('button'));
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('does render default accessibility attributes correctly', () => {
    const { container } = renderWithProviders(
      <SMLogo onClick={() => {}} />,
    );
    const buttonBase = container.querySelector('button');
    // Expect role to be link
    expect(buttonBase).toHaveAttribute('role', 'link');
    // Since we are mocking home path for location aria-current should be page
    expect(buttonBase).toHaveAttribute('aria-current', 'page');
    // Home logo which is image should containe image text in aria label
    // so aria label should contain "Palvelukartta" text in addition to action description
    expect(buttonBase).toHaveAttribute('aria-label', englishTranslations['general.home.logo.ariaLabel']);
    // Expect visible text to be hidden from screen readers
    expect(buttonBase.querySelector('div')).toHaveAttribute('aria-hidden', 'true');
    expect(buttonBase.querySelector('div')).toHaveAttribute('role', 'img');
  });
});
