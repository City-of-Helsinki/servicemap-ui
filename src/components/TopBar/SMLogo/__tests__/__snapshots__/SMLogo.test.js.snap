// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SMLogo /> should work 1`] = `
<div>
  <button
    aria-current="page"
    aria-label="Servicemap - Go to home page"
    class="MuiButtonBase-root css-fvdnvy-MuiButtonBase-root"
    role="link"
    tabindex="0"
    type="button"
  >
    <div
      aria-hidden="true"
      class="SMLogoComponent-logo-6"
      role="img"
    >
      <img
        alt=""
        class="ForwardRef-icon-27"
        src="test-file-stub"
      />
    </div>
  </button>
</div>
`;
